x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DelayedDataDetector.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.datafeed;
+
+import org.elasticsearch.action.search.SearchAction;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.histogram.Histogram;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xpack.core.ml.action.GetBucketsAction;
+import org.elasticsearch.xpack.core.ml.action.util.PageParams;
+import org.elasticsearch.xpack.core.ml.datafeed.DatafeedConfig;
+import org.elasticsearch.xpack.core.ml.datafeed.extractor.ExtractorUtils;
+import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.core.ml.job.results.Bucket;
+import org.elasticsearch.xpack.core.ml.utils.Intervals;
+import org.joda.time.DateTime;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.stashWithOrigin;
+
+
+/**
+ * This class will search the buckets and indices over a given window to determine if any data is missing
+ */
+public class DelayedDataDetector {
+
+    private static final String DATE_BUCKETS = "date_buckets";
+    private final long bucketSpan;
+    private final long window;
+    private final DatafeedConfig datafeedConfig;
+    private final Client client;
+    private final Job job;
+
+    public DelayedDataDetector(Job job, DatafeedConfig datafeedConfig, TimeValue window, Client client) {
+        this.job = job;
+        this.bucketSpan = job.getAnalysisConfig().getBucketSpan().millis();
+        this.datafeedConfig = datafeedConfig;
+        long windowMillis = window.millis();
+        if (windowMillis < bucketSpan) {
+            throw new IllegalArgumentException("[window] must be greater or equal to the [bucket_span]");
+        }
+        this.window = windowMillis;
+        this.client = client;
+    }
+
+    public List<BucketWithMissingData> missingData(long latestFinalizedBucket) {
```

Comment:
It would be good if there was a verb in the method name to indicate that it does significant work, so maybe `detectMissingData` given that the class is a "Detector".

Also, since it's a `public` method it would be good to have a Doxygen comment saying that the method may block for a considerable time and hence must be run in an appropriate thread pool.